Source/Renderer/ShaderProgram.js

```diff
@@ -27,7 +29,51 @@ define([
     /**
      * @private
      */
+
+    function extractUniforms(shaderText) {
+        // TODO: regex doesn't work for inline structs
+        var uniformNames = [];
+        var uniformLines = shaderText.match(/uniform[\w\s]*(?=[=\[;])/g);
+        if (uniformLines) {
+            for (var i = 0; i < uniformLines.length; i++) {
+                var line = uniformLines[i].trim();
+                var name = line.slice(line.lastIndexOf(' ') + 1);
+                uniformNames.push(name);
+            }
+        }
+        return uniformNames;
+    }
+
     var ShaderProgram = function(options) {
+        // If a uniform exists in both the vertex and fragment shader but with different precision qualifiers,
+        // give the fragment shader uniform a different name. This fixes shader compilation errors on devices
+        // that only support mediump in the fragment shader.
+        var vertexShaderText = options.vertexShaderText;
+        var fragmentShaderText = options.fragmentShaderText;
+        var duplicateUniformNames = {};
+
+        if (!ContextLimits.highpSupported) {
```

Comment:
We generally want all the code in a function to be at the same level of abstraction.  I would probably break this logic into its own function that returns an object with two properties: `fragmentShaderText` and `duplicateUniformNames`.